### PR TITLE
feat: add small pieces of max theory

### PR DESCRIPTION
**Pull request by RaitoBezarius**
`0#ty` is neutral for `max` for unsigned integers.

### CODE REVIEW

File: backends/lean/Base/Primitives/Scalar.lean [hunk 1]
```diff
@@ -1419,6 +1423,21 @@ instance (ty: ScalarTy) : LinearOrder (Scalar ty) where
...
+
+-- Max theory
+-- TODO: do the min theory later on.
+
+theorem Scalar.zero_le_unsigned {ty} (s: ¬ ty.isSigned) (x: Scalar ty): Scalar.ofInt 0 ≤ x := by
+  apply (Scalar.le_equiv _ _).2
+  convert x.hmin
+  cases ty <;> simp [ScalarTy.isSigned] at s <;> simp [Scalar.min]
+
+theorem Scalar.max_unsigned_left_zero_eq {ty} (s: ¬ ty.isSigned) (x: Scalar ty):
```
Comment (sonmarcho) [inline backends/lean/Base/Primitives/Scalar.lean:1435]: Do you confirm that simplification with this theorem works? In particular, is the `¬ ty.isSigned` automatically discharged when you expect it to be discharged?

Comment (RaitoBezarius) [reply backends/lean/Base/Primitives/Scalar.lean:1435]: No, I think we determined it doesn't, it must be provided manually.

Comment (sonmarcho) [reply backends/lean/Base/Primitives/Scalar.lean:1435]: I think we had figured out a potential solution, by using typeclasses?

Comment (RaitoBezarius) [reply backends/lean/Base/Primitives/Scalar.lean:1435]: Yep, we need to use the `Fact` typeclass to have that predicate around.

Comment (RaitoBezarius) [reply backends/lean/Base/Primitives/Scalar.lean]: @sonmarcho I pushed a version with the fact typeclass, should we do the fact instances in a upcoming PR? I am not exactly sure how we want to do them, e.g. just copy/paste `instance : Fact (not ty.isSigned) where out := ...` for all `ty` or just do a bit of metaprog to generate all those instances with a similar tactic?

Comment (sonmarcho) [reply backends/lean/Base/Primitives/Scalar.lean]: I'd rather do them in this PR: I want to get those simp lemmas working before merging anything. Having one instance per unsigned integer type would work.

Comment (RaitoBezarius) [reply backends/lean/Base/Primitives/Scalar.lean]: OK, I can do that

Comment (RaitoBezarius) [reply backends/lean/Base/Primitives/Scalar.lean]: This is done, @sonmarcho!

Comment (sonmarcho) [reply backends/lean/Base/Primitives/Scalar.lean]: Excellent, thanks! :)

**Comment by sonmarcho** [comment]
There are conflicts with `main` now it seems

**Comment by RaitoBezarius** [comment]
I will rebase.

**Pull request merged**